subsys/net/ip/net_shell.c

```diff
@@ -3534,159 +3753,249 @@ static void iface_vlan_cb(struct net_if *iface, void *user_data)
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_gptp)
+{
+	SHELL_CMD(port, NULL, "Print detailed information about gPTP port.",
+		  cmd_net_gptp_port),
+	SHELL_SUBCMD_SET_END
 };
 
-SHELL_REGISTER(NET_SHELL_MODULE, net_commands);
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_http)
+{
+	SHELL_CMD(monitor, NULL, "Start monitoring HTTP connections.",
+		  cmd_net_http_monitor),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_iface)
+{
+	SHELL_CMD(up, NULL, "Take network interface up.", cmd_net_iface_up),
+	SHELL_CMD(down, NULL, "Take network interface down.",
+		  cmd_net_iface_down),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_nbr)
+{
+	SHELL_CMD(rm, NULL, "Remove neighbor from cache.",
+		  cmd_net_nbr_rm),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_stats)
+{
+	SHELL_CMD(all, NULL,
+		  "Show network statistics for all network interfaces.",
+		  cmd_net_stats_all),
+	SHELL_CMD(iface, NULL,
+		  "Show network statistics for one network interface.",
+		  cmd_net_stats_iface),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_tcp)
+{
+	SHELL_CMD(connect, NULL, "Connect to TCP peer.", cmd_net_tcp_connect),
+	SHELL_CMD(send, NULL, "Send data to peer using TCP.",
+		  cmd_net_tcp_send),
+	SHELL_CMD(close, NULL, "Close TCP connection.", cmd_net_tcp_close),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_vlan)
+{
+	SHELL_CMD(add, NULL, "Add VLAN tag to the network interface.",
+		  cmd_net_vlan_add),
+	SHELL_CMD(del, NULL, "Delete VLAN tag from the network interface.",
+		  cmd_net_vlan_del),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_commands)
+{
+	/* Alphabetically sorted. */
+	SHELL_CMD(allocs, NULL, "Print network memory allocations.",
+		  cmd_net_allocs),
+	SHELL_CMD(app, NULL,
+		  "Print network application API usage information.",
+		  cmd_net_app),
+	SHELL_CMD(arp, &net_cmd_arp, "Print information about IPv4 ARP cache.",
+		  cmd_net_arp),
+	SHELL_CMD(conn, NULL, "Print information about network connections.",
+		  cmd_net_conn),
+	SHELL_CMD(dns, &net_cmd_dns, "Show how DNS is configured.",
+		  cmd_net_dns),
+	SHELL_CMD(gptp, &net_cmd_gptp, "Print information about gPTP support.",
+		  cmd_net_gptp),
+	SHELL_CMD(http, &net_cmd_http,
+		  "Print information about active HTTP connections.",
+		  cmd_net_http),
+	SHELL_CMD(iface, &net_cmd_iface,
+		  "Print information about network interfaces.",
+		  cmd_net_iface),
+	SHELL_CMD(mem, NULL, "Print information about network memory usage.",
+		  cmd_net_mem),
+	SHELL_CMD(nbr, &net_cmd_nbr, "Print neighbor information.",
+		  cmd_net_nbr),
+	SHELL_CMD(ping, NULL, "Ping a network host.", cmd_net_ping),
+	SHELL_CMD(route, NULL, "Show network route.", cmd_net_route),
+	SHELL_CMD(rpl, NULL, "Show RPL mesh routing status.", cmd_net_rpl),
+	SHELL_CMD(stacks, NULL, "Show network stacks information.",
+		  cmd_net_stacks),
+	SHELL_CMD(stats, &net_cmd_stats, "Show network statistics.",
```

Comment:
You could solve 'stats' command in more handy way by utilizing dynamic autocompletion. It could have syntax like:
- _net stats_ prints stats for all interfaces
- _net_stats <iface1> <iface2>_ - prints stats for all named interfaces
Very similar approach is taken in _log disable_ command (https://github.com/zephyrproject-rtos/zephyr/blob/f2596952599df87d9dafaa8d692f7f737a0fb3f6/subsys/logging/log_cmds.c#L390)

Reply:
That would be cool, there are also other commands that could benefit from this.

Reply:
I have been experimenting with the dynamic commands. I added dynamic completion for `net iface up <idx>` command. I got it working but it is not very user friendly atm.
In this example I have two network interfaces. When completing the command I see this:
```
net:~$ net iface up <tab>
  0  1
```
but it is not easy to understand what the values actually mean. For that one needs to add --help string in which case I get
```
net:~$ net iface up 1 --help
up - 'net iface up <iface index>' takes network interface up.
Options:
  -h, --help  :Show command help.
Subcommands:
  0  :Ethernet (0x00421860)
  1  :Ethernet (0x00421920)
```

For user this is quite cumbersome. Why not show the help during completion like this:
```
net:~$ net iface up <tab>
  0  :Ethernet (0x00421860)
  1  :Ethernet (0x00421920)
```

This would be much more user friendly and user could select the proper index immediately. If there is only one interface, then it is selected automatically currently and this is just fine.

Reply:
So you would like to have help string for promted dynamic commands or all commands?
Or maybe I could prepare special type of dynamic commands printing this help?

Reply:
I was just setting `entry->help  = set_iface_index_help(idx);` for dynamic command, and was expecting that it's value would be used in completion. Will send new version shortly so you can see how I did this.

Reply:
Setting `entry->help` only allows to print help when command is called with -h or --help.
Otherwise it would be inconsistency between static and dynamic commands.

When you press tab you only see command syntax, not command syntax + help.
Maybe you could change a dynamic command syntax name from `0` to `0 :Ethernet (0x00421860)` ?

Reply:
> Maybe you could change a dynamic command syntax name from 0 to 0 :Ethernet (0x00421860) ?

I tried that too as that seemed like a logical thing to do. So after pressing <tab>, the "0 :Ethernet (0x00421860)" and "1: Ethernet <(0xabcd01234)" was printed, and by selecting 0 and then pressing <tab> again the "0 :Ethernet (0x00421860)" is selected which is of course wrong. Naturally one would probably not press <tab> after selecting 0 but the point is that the expansion goes wrong in this case.
So I was just wondering if the "entry->help" string could be used here to guide user what the options means when pressing <tab>

Reply:
I am not quite sure I understand you. Please correct me if I am wrong:
1. You have changed dynamic command syntax from `0` to `0 :Ethernet (0x00421860)` and from `1` to `1: Ethernet <(0xabcd01234)`
2. When you press Tab button you can see prompted 2 options: `0 :Ethernet (0x00421860)` and "1: Ethernet <(0xabcd01234)".
3. You are pressing `0` next Tab and shell has autocomplete `0` to  `0 :Ethernet (0x00421860)` and you did not expect it?

Reply:
To be honest I do not see easy solution for now to prompt command syntax + command help but to autocompleate only command name. :(

Reply:
I will probably remove the dynamic completion support from net-shell as the feature uses lot of memory. Especially the IPv6 neighbor completion uses 240 bytes memory by default. The network interface index completion is using smaller amount of memory but is currently not very user friendly.

Reply:
> Could we use interface names, like eth0, at shell level?

Currently we do not have a concept for interface name in IP stack, if we introduce such a name it would be only for shell which complicates things. This needs to be pondered more, atm there are more important things to do so this is left for later (at least from my part). Patches are welcome of course.

> if we make this option configurable.

Yes, I was planning to make the completion configurable and enable it by default.